### PR TITLE
Ensure the shared HashStringAllocator isn't mutated by Aggregations during spilling

### PR DESCRIPTION
**Pull request by kevinwilfong**
Summary:
In https://github.com/facebookincubator/velox/pull/10268 we discovered and fixed an issue where
approx_percentile was running into concurrency issues in HashStringAllocator during spilling
because parallel spiling would lead to multiple concurrent allocations from the same
HashStringAllocator.

This was fixed for approx_percentile, but there may still be issues in other aggregates (existing or
future).  To catch these clearly and early I've add a mechanism to freeze HashStringAllocator.

When HashStringAllocator is frozen any attempt to access any of its fields in a mutable way will
result in an exception.  This is achieved by wrapping its fields in an inner class State which exposes
them via getter methods that can either return a const or non-const reference.  Before returning a
non-const reference it first checks if the class is frozen.

I've updated the places in GroupingSet where we invoke spilling to freeze the HashStringAllocator
while this process is happening.  This way if any Aggregation may trigger a concurrency issue it is
detected and an exception is thrown.

This relies on future developers following the pattern of adding fields to the State class with
accessors, I've added comments to do my best to ensure they follow this pattern.

Differential Revision: D59012548

### CODE REVIEW

File: velox/common/memory/HashStringAllocator.h [hunk 1]
```diff
@@ -338,6 +340,16 @@ class HashStringAllocator : public StreamArena {
 
   std::string toString() const;
 
+  /// Effectively makes this immutable while executing f, any attempt to access
+  /// state_ in a mutable way while f is executing will cause an exception to be
+  /// thrown.
+  template <typename F>
+  void freezeAndExecute(F&& f) {
+    state_.freeze();
+    auto guard = folly::makeGuard([&] { state_.unfreeze(); });
```
Comment (xiaoxmeng) [inline velox/common/memory/HashStringAllocator.h:349]: We can use SCOPE_EXIT here

File: velox/common/memory/HashStringAllocator.h [hunk 2]
```diff
@@ -392,38 +404,108 @@ class HashStringAllocator : public StreamArena {
...
+  /// A class that wraps any fields in the HashStringAllocator, it's main
+  /// purpose is to simplify the freeze/unfreeze mechanic.  Fields are exposed
+  /// via accessor methods, attempting to invoke a non-const accessor when the
+  /// HashStringAllocator is frozen will cause an exception to be thrown.
+  class State {
+    typedef CompactDoubleList FreeList[kNumFreeLists];
+    typedef uint64_t FreeNonEmptyBitMap[bits::nwords(kNumFreeLists)];
+    typedef folly::F14FastMap<void*, size_t> AllocationsFromPool;
+
+   public:
```
Comment (xiaoxmeng) [inline velox/common/memory/HashStringAllocator.h:416]: Put public first?

File: velox/common/memory/HashStringAllocator.h [hunk 3]
```diff
@@ -392,38 +404,108 @@ class HashStringAllocator : public StreamArena {
...
+    explicit State(memory::MemoryPool* pool) : pool_(pool) {}
+
+    void freeze() {
+      VELOX_CHECK(
+          isMutable_,
```
Comment (xiaoxmeng) [inline velox/common/memory/HashStringAllocator.h:421]: nit: s/isMutable_/mutable_/

File: velox/common/memory/HashStringAllocator.h [hunk 4]
```diff
@@ -392,38 +404,108 @@ class HashStringAllocator : public StreamArena {
...
+
+// Declare a default initialized field.
+#define DECLARE_FIELD(TYPE, NAME) \
+  DECLARE_GETTERS(TYPE, NAME)     \
+                                  \
+ private:                         \
+  TYPE NAME##_;
+
+// Declare a field initialized with a specific value.
+#define DECLARE_FIELD_INIT_VALUE(TYPE, NAME, VALUE) \
```
Comment (xiaoxmeng) [inline velox/common/memory/HashStringAllocator.h:455]:
```
nit: s/DECLARE_FIELD_INIT_VALUE/DECLARE_FIELD_WITH_INIT_VALUE/
```

File: velox/exec/GroupingSet.cpp [hunk 5]
```diff
@@ -962,8 +962,8 @@ void GroupingSet::spill() {
     return;
   }
 
+  auto rows = table_->rows();
```
Comment (xiaoxmeng) [inline velox/exec/GroupingSet.cpp:965]:
```
s/auto rows/auto* rows/
```

File: velox/common/memory/tests/HashStringAllocatorTest.cpp [hunk 6]
```diff
@@ -678,5 +678,31 @@ TEST_F(HashStringAllocatorTest, clear) {
...
+  std::string str = "abc";
+  StringView view(str.data(), str.size());
+  allocator_->copyMultipart(view, reinterpret_cast<char*>(&view), 0);
+
+  str.clear();
+
+  // Freeing memory requires the HashStringAllocator to be mutable.
+  VELOX_ASSERT_THROW(
+      allocator_->freezeAndExecute([&]() {
+        allocator_->free(HashStringAllocator::headerOf(view.data()));
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/HashStringAllocatorTest.cpp:691]: Can you add a test that const method access is ok? Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@kevinwilfong nice change. Thanks!

File: velox/common/memory/tests/HashStringAllocatorTest.cpp (hunk 6 above)
Comment (kevinwilfong) [reply velox/common/memory/tests/HashStringAllocatorTest.cpp:691]: That was the intention of calling cumulativeBytes() at the bottom of this test.

**Pull request closed without merging**